Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -1186,9 +1214,67 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     m_backgroundWebView = nil;
 }
 
+WebExtensionContext::EventListenterTypeCountedSet WebExtensionContext::loadBackgroundPageListenersFromStorage()
+{
+    NSDictionary *extensionState = currentState();
+    NSCountedSet *savedListeners = [NSKeyedUnarchiver unarchivedObjectOfClasses:[NSSet setWithArray:@[ NSCountedSet.class, NSNumber.class ]] fromData:[extensionState objectForKey:backgroundPageListenersKey] error:nil];
+
+    EventListenterTypeCountedSet result;
+    for (NSNumber *entry in savedListeners)
+        result.add(static_cast<WebExtensionEventListenerType>(entry.unsignedIntValue), [savedListeners countForObject:entry]);
+
+    return result;
+}
+
+NSString *WebExtensionContext::lastSeenUniqueIdentifier()
+{
+    return [m_state objectForKey:lastSeenUniqueIdentifierKey];
+}
+
+void WebExtensionContext::saveBackgroundPageListenersToStorage()
+{
+    if (backgroundContentURL().isEmpty() || isPersistent())
+        return;
+
+    NSMutableDictionary *extensionState = [currentState() mutableCopy];
+    if (!extensionState.count)
+        return;
+
+    NSCountedSet *listeners = [NSCountedSet set];
+    for (auto& entry : m_backgroundPageListeners)
+        [listeners addObject:@(static_cast<uint8_t>(entry.key))];
+
+    NSData *newBackgroundPageListenersAsData = [NSKeyedArchiver archivedDataWithRootObject:listeners requiringSecureCoding:YES error:nil];
+    NSData *savedBackgroundPageListenersAsData = [extensionState objectForKey:backgroundPageListenersKey];
+    extensionState[backgroundPageListenersKey] = newBackgroundPageListenersAsData;
+
+    NSNumber *savedListenerVersionNumber = [extensionState objectForKey:lastSeenVersionStateKey];
+    extensionState[lastSeenVersionStateKey] = @(currentBackgroundPageListenerStateVersion);
+
+    BOOL hasListenerStateChanged = ![newBackgroundPageListenersAsData isEqualToData:savedBackgroundPageListenersAsData];
+    BOOL hasVersionNumberChanged = savedListenerVersionNumber.integerValue != currentBackgroundPageListenerStateVersion;
+    if (hasListenerStateChanged || hasVersionNumberChanged) {
+        m_state = extensionState;
+        writeStateToStorage();
+    }
+}
+
+void WebExtensionContext::saveUniqueIdentifierToStorage()
+{
+    [m_state setObject:m_uniqueIdentifier forKey:lastSeenUniqueIdentifierKey];
+    writeStateToStorage();
+}
+
+NSNumber *WebExtensionContext::loadBackgroundPageListenersVersionNumberFromStorage()
```

Comment:
Instead of returning an `NSNumber` return a C type like `NSInteger`. It lets the client avoid the unboxing.

Reply:
Should still do this. If you hit an error with `NSInteger` try `uint64_t`.